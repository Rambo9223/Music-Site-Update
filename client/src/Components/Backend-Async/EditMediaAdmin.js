/* PUT Function to edit a media items title, path or link
 */
const EditMediaAdmin = async(item,token) => {
    // item to edit & auth token
    const response = await fetch("/admin/media", {
       method: "PUT",
       headers: {
         "Content-Type": "application/json",
         token:token
       },
       body:JSON.stringify(item),
     });
     // return response.json
     if (response.status !== 200) {
       return response.json();
     } else {
       return response.json();
     }
   };

   export default EditMediaAdmin