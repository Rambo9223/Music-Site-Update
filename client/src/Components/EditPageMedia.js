import { useForm } from "react-hook-form";
import {Modal,Button} from "react-bootstrap";
import { XCircle,CheckCircle } from "react-bootstrap-icons";
import EditMediaAdmin from "./Backend-Async/EditMediaAdmin";
import PopUp from "./PopUp";
import { useState } from "react";

/* Page that allows admins to edit details on the page media items */

export default function EditPageMedia(props){
    let item = props.item;// media item 
    let token = props.token// auth token to pass to server
    let bool = props.bool;// bool for modal 
    let ToggleBool = props.ToggleBool;// function to open/close modal

    let initialPopUp = {// sucess or failure popup 
        message:"",
        type:"",
        response:""
      }
    
    // bool for response popup 
    const [resBool,setResBool] = useState(false);
    // function to add details to popup object
    const [popup,setPopUp] = useState(initialPopUp)
    // toggle response popup and close edit media popup
    const ToggleResBool = () => {setResBool(!resBool);
    if(resBool===true){ToggleBool();}}

    // inintalise react hook form
    const {
        register,
        handleSubmit,
        formState: { errors },
      } = useForm({
        defaultValues:{
            "title":item.title,
            "path":item.path,
            "link":item.link
        }
      });

    // function to handle form submit
    const onSubmit = (data) => {
        EditMediaAdmin(data,token).then((res)=>{// send to backend 
            if(res.ok===true){// if respose is good
                // open response popup with black text
                setResBool(true);
                setPopUp({message:res.message,type:"Sever Response",response:"text-primary"})
            }else{
                // else open response popup with red text
                setResBool(true);
                setPopUp({message:res.message,type:"Sever Response",response:"text-danger"})
            }
        })
        
    }

    return <div>
    {(resBool===false)?
    <Modal show={bool} onHide={ToggleBool}>
        <Modal.Header closeButton>
        <Modal.Title>
          <h2 className="text-dark">Edit Media</h2>
        </Modal.Title>
        </Modal.Header>
        <Modal.Body>
        <form className="new-item-form" onSubmit={handleSubmit(onSubmit)}>
          <label className="form-labels">Title:</label>
          <br />
          <input
            className="form-inputs"
            type="text"
            {...register("title", {
              required: "Title is required.",
            })}
          />
          {errors.title && <p className="errorMsg">{errors.title.message}</p>}
          <br />
          <label className="form-labels">Path: </label>
          <input
            className="form-inputs"
            type="text"
            name="path"
            {...register("path", {
              required: "Path is required.",
            })}
          />
          {errors.path && <p className="errorMsg">{errors.path.message}</p>}
          <br />
          {(item.link)?<><label className="form-labels">Link: </label>
          <input
            className="form-inputs"
            type="text"
            {...register("link", {
              required: "Link is required.",
            })}
          />
          {errors.link && <p className="errorMsg">{errors.link.message}</p>}
          <br /></>:null}
          <label className="form-labels">Id:</label>
          <input
            className="form-inputs"
            type="text"
            value={item.id}
            name="id"
            {...register("id")}
          />
          <label className="form-labels">Media Type:</label>
          <input
            className="form-inputs"
            type="text"
            value={item.media}
            name="media"
            {...register("media")}
          />
        <Button variant="primary" onClick={ToggleBool}>Close <XCircle/>
        </Button>
        <Button variant="success" type="submit">Edit <CheckCircle/></Button>
        </form>
        </Modal.Body>
        <Modal.Footer>
        </Modal.Footer>
    </Modal>
    :<PopUp message={popup.message} type={popup.type} response={popup.response} bool={resBool} ToggleBool={ToggleResBool}/>}
    </div>

}