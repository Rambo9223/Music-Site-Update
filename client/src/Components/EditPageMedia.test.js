/**
*@jest-environment jsdom 
*/
import { render,cleanup, waitFor} from '@testing-library/react';
import EditPageMedia from './EditPageMedia';
import EditMediaAdmin from './Backend-Async/EditMediaAdmin';
import * as React from 'react';


afterEach(()=>{
  cleanup();
  jest.clearAllMocks();
});


describe('Edit Page Media Tests', () => { 


  let bool = true;// variable to open/close modal
    function changeBool(){// simulates usestate ToggleBool
      bool = !bool;
    }
    let item = {
        "title":"test title",
        "path":"test path",
        "link":"test link"
    }
  
  // snapshot test
  test('Page matches snapshot', () => {
    let page = render(<EditPageMedia item={item} token={"string"} bool={true} ToggleBool={changeBool}/>);
    expect(page).toMatchSnapshot();
  });

  // form exists when modal is visible
  test('Form opens in modal when bool is true', async () => {
    // render page with props
    let page = render(<EditPageMedia  item={item} token={"string"} bool={bool} ToggleBool={changeBool}/>)

    // on bool = true form should be visible
    let form = (await page.findByText("Edit Media"));

    // expect form to be in document
    await waitFor(()=>{expect(form).toBeInTheDocument();}).finally(()=>{
      changeBool();// simulates a change in usestate ToogleBool for closing the form for next test
    });
  })

  // check form closes when modal is closed
  test('Form & Modal closes when bool is false', async () => { 

    // page is rendered with bool = false
    let page = render(<EditPageMedia item={item} token={"string"} bool={bool} ToggleBool={changeBool}/>)

    // label is a html element that is within the form
    let label = page.queryByText("Media Type:");// use query as we expect returned answer to be null 

    // final assertion
    expect(label).not.toBeInTheDocument()

   });

   test('Submit mock item on form submit', async() => {
    let res = {
        status:200,
        ok:true,
        message:`Media item: ${"test title"} has been successfully edited.`
    }
    global.fetch = jest.fn().mockImplementationOnce(
        () => new Promise((resolve) => {
            resolve({
                // first, mock the "json" function containing our result
                json: () => new Promise((resolve) => {
                  // then, resolve this second promise with our object
                  resolve(res)
                }),})
})); 

    let page = render(<EditPageMedia item={item} token={"string"} bool={true} ToggleBool={changeBool}/>);

    let button = await page.findByText("Edit");

    expect(button).toBeInTheDocument();
    // button exists so imagine we mock a submit of the form

    // we would then call fetch on the async function
    let result = await EditMediaAdmin(item,"token");
    // expect fetch to be called once
    expect(fetch).toHaveBeenCalledTimes(1);
    // result is equal to expected
    expect(res).toEqual(result);

   })


 })