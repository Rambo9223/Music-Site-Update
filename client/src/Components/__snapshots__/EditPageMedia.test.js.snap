// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Page Media Tests Page matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    data-rr-ui-modal-open=""
    style="padding-right: 1024px;"
  >
    <div>
      <div />
    </div>
    <div
      class="fade modal-backdrop show"
    />
    <div
      aria-modal="true"
      class="fade modal show"
      role="dialog"
      style="display: block; padding-right: 0px;"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <div
              class="modal-title h4"
            >
              <h2
                class="text-dark"
              >
                Edit Media
              </h2>
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <form
              class="new-item-form"
            >
              <label
                class="form-labels"
              >
                Title:
              </label>
              <br />
              <input
                class="form-inputs"
                name="title"
                type="text"
              />
              <br />
              <label
                class="form-labels"
              >
                Path: 
              </label>
              <input
                class="form-inputs"
                name="path"
                type="text"
              />
              <br />
              <label
                class="form-labels"
              >
                Link: 
              </label>
              <input
                class="form-inputs"
                name="link"
                type="text"
              />
              <br />
              <label
                class="form-labels"
              >
                Id:
              </label>
              <input
                class="form-inputs"
                name="id"
                type="text"
                value=""
              />
              <label
                class="form-labels"
              >
                Media Type:
              </label>
              <input
                class="form-inputs"
                name="media"
                type="text"
                value=""
              />
              <button
                class="btn btn-primary"
                type="button"
              >
                Close 
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14m0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16"
                  />
                  <path
                    d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"
                  />
                </svg>
              </button>
              <button
                class="btn btn-success"
                type="submit"
              >
                Edit 
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14m0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16"
                  />
                  <path
                    d="m10.97 4.97-.02.022-3.473 4.425-2.093-2.094a.75.75 0 0 0-1.06 1.06L6.97 11.03a.75.75 0 0 0 1.079-.02l3.992-4.99a.75.75 0 0 0-1.071-1.05"
                  />
                </svg>
              </button>
            </form>
          </div>
          <div
            class="modal-footer"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
